Add "How to capture client-side errors on a page" (#24)
DevExpress/testcafe-examples: dedicated-examples/extended-error-tracking/README.MD

> # Log Client-Side Errors

Reviewer: `Extended Client-Side error tracking`.

When a JavaScript error occurs on a tested web page, TestCafe stops test execution and posts an error message. Requests to foreign resources (e.g. third-party APIs) might cause unexpected errors that are not relevant to your test. In that case, you might want to filter those errors out to ignore them during the test.

> This example demonstrates how to expand on TestCafe's standard error logging and filter client-side errors that happen during a test.

Reviewer: `This example demonstrates how to expand`. `extend` is a better word for this place.

Author: > Third-party code execution (e.g. logging and analytics services, controls) might cause unexpected errors that are not relevant to the tested app. You might want to filter those errors out to ignore them during the test.

**Test Code**: [index.js](index.js)

The test page has 2 buttons, each throws an error with `.level` and `.url` custom properties when clicked:

||level|url|
|------|--------|-----|
|`Error 1`|`critical`|`https://my-domain.example.com`|
|`Error 2`|`warning`|`https://foreign-domain.example.com`|

These properties are later used to filter these errors.

> During the test, both buttons receive clicks. To log the errors, [test.clientScripts](https://devexpress.github.io/testcafe/documentation/reference/test-api/test/clientscripts.html) method [injects a client script](https://devexpress.github.io/testcafe/documentation/guides/advanced-guides/inject-client-scripts.html) into the page. This script records all the errors that happen during the test into an array.

Reviewer: `This script collects all the errors...`

Author: > This script collects all the errors that happen during the test and records them into an array.

> That array is then checked with the [Array.some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) JavaScript method to find whether any `critical` errors have happened on the `https://my-domain.example.com` domain.

Reviewer: > have happened on the `https://my-domain.example.com` domain.

It's wrong. These errors occurred on the current domain. `url` is just a property.

That array is then checked with the [Array.some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) JavaScript method to find whether any `critical` errors have happened on the tested web page.

Author: > The array is then checked with the [Array.some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) JavaScript method to find whether any `critical` errors have happened on the tested web page. If such errors have occurred, the test fails, otherwise, it passes.

> ## Launch Configuration

Reviewer: At present, this test is not executed on CI.

Add a new item to the `scripts` section and call it in the `test` script.

```js
"scripts": {
    "examples": "testcafe chrome,firefox examples -a \"node server/index.js\"",
    "extended-error-tracking": "testcafe chrome dedicated-examples/extended-error-tracking/index.js -e",
    "mock-camera-microphone-access": " testcafe --hostname localhost \"chrome --use-fake-ui-for-media-stream --use-fake-device-for-media-stream\" dedicated-examples/mock-camera-microphone-access/mock-camera-microphone-access.js",
    "test": "npm run examples && npm run mock-camera-microphone-access && extended-error-tracking"
  },
```

Reviewer: ## Launch Configuration

When a JavaScript error occurs on a tested web page, TestCafe stops test execution. To prevent this, the `-e` [CLI option](https://devexpress.github.io/testcafe/documentation/reference/command-line-interface.html#-e---skip-js-errors) flag is used.

To launch the only this example, use the following command:
```sh
npm run extended-error-tracking;
```

Author: > When a JavaScript error occurs on a tested web page, TestCafe stops test execution. To prevent this, use the `-e` [CLI option](https://devexpress.github.io/testcafe/documentation/reference/command-line-interface.html#-e---skip-js-errors).

To launch this example directly on its own, use the following command:

```sh
npm run extended-error-tracking;
```

# Extended Client-Side Error Tracking

> When a JavaScript error occurs on a tested web page, TestCafe stops test execution and posts an error message. Third-party code execution (e.g. controls, logging and analytics services) might cause unexpected errors that are not relevant to the tested app. You might want to filter those errors out to ignore them during the test.

Reviewer 2: ```suggestion
When a JavaScript error occurs on a tested web page, TestCafe stops test execution and posts an error message. Third-party code execution (e.g., controls, logging, and analytics services) may cause unexpected errors that are not relevant to the tested app. You might want to filter those errors out to ignore them during the test.
```

Reviewer 2: You might want to filter those errors out to ignore them during the test.
This sentence doesn't make sense. Can you rewrite?

Author: > When a JavaScript error occurs on a tested web page, TestCafe stops test execution and posts an error message. Third-party code execution (for example, controls, logging, and analytics services) can cause unexpected errors that are not related to the tested app. You can filter these errors to ignore them during the test.

Reviewer 2: You can filter these errors so that they are not included in the test.

Author: edited

> When a JavaScript error occurs on a tested web page, TestCafe stops test execution and posts an error message. Third-party code execution (for example, controls, logging, and analytics services) can cause unexpected errors that are not related to the tested app. You can filter these errors so that they are not included in the test.

Reviewer 2: ```suggestion
When a JavaScript error occurs on a tested web page, TestCafe stops test execution and posts an error message. Third-party code execution (for example, controls, logging, and analytics services) can cause unexpected errors that are not related to the tested app. You can filter these errors, so that they are not included in the test.
```

> When a JavaScript error occurs on a tested web page, TestCafe stops test execution and posts an error message to the console. Third-party code execution (e.g. controls, logging and analytics services) might cause unexpected errors that are not relevant to the tested app. You might want to filter those errors out to ignore them during the test.

Reviewer: > an error message to the console.

TestCafe can report the test execution log to any output: console, file, REST API call and etc.

Let's remove `to the console` phrase.

Author: done

> This example demonstrates how to extend TestCafe's standard error tracking and filter client-side errors that occur during a test.

Reviewer 2: ```suggestion
The following example demonstrates how to extend TestCafe's standard error tracking and how to filter client-side errors that occur during a test.
```

> This example demonstrates how to extend TestCafe's standard error tracking and filter client-side errors that happen during a test.

Reviewer 3: Errors more often "occur" than "happen"

Author: agreed, edited

> This example demonstrates how to expand on TestCafe's standard error logging and filter client-side errors that happen during a test.

Reviewer: `This example demonstrates how to extend TestCafe's standard error tracking and filter...`

Author: done

**Test Code**: [index.js](index.js)

> The test page has 2 buttons (`Error 1` and `Error 2`), which throw an error with `.level` and `.url` custom properties when clicked. These properties are used to filter errors and have the following values:

Reviewer 2: ```suggestion
The test page has two buttons (`Error 1` and `Error 2`). When clicked, these buttons throw an error with `.level` and `.url` custom properties that are used to filter errors and have the following values:
```

> The test page has 2 buttons (`Error 1` and `Error 2`), each throws an error with `.level` and `.url` custom properties when clicked. These properties are used to filter the errors and have the following values:

Reviewer 2: ```suggestion
The test page has 2 buttons (`Error 1` and `Error 2`), which throw an error with `.level` and `.url` custom properties when clicked. These properties are used to filter errors and have the following values:
```

> The test page has 2 buttons (`Error 1` and `Error 2`), each throws an error with `.level` and `.url` custom properties when clicked. These properties are later used to filter the errors and have the following values:

Reviewer 3: I think here we don't have to say 'later' - the properties are obviously used after they are defined

|Button|.level|.url|
|------|--------|-----|
|`Error 1`|`critical`|`https://my-domain.example.com`|
|`Error 2`|`warning`|`https://foreign-domain.example.com`|

> During the test, both buttons are clicked. To capture the errors, [test.clientScripts](https://devexpress.github.io/testcafe/documentation/reference/test-api/test/clientscripts.html) method injects a client script into the page. This script collects all the errors that occur during the test and records them into an array.

Reviewer 2: ```suggestion
During the test, both buttons are clicked. To capture the errors, the [test.clientScripts](https://devexpress.github.io/testcafe/documentation/reference/test-api/test/clientscripts.html) method injects a client script into the page. This script collects all the errors that occur during the test and records them into an array.
```

> During the test, both buttons receive clicks. To capture the errors, [test.clientScripts](https://devexpress.github.io/testcafe/documentation/reference/test-api/test/clientscripts.html) method injects a client script into the page. This script collects all the errors that happen during the test and records them into an array.

Reviewer 3: Controls can "receive focus" but I don't usually come across "receive clicks". `During the test, both buttons are clicked.`?

I suspect the pattern we established (`During the test, ...`) dictates some weird phrasing. But it doesn't have to, we can use it with simple common phrases.

Reviewer 3: And `errors that occur` here as well

Reviewer 3: I could say "mistakes happen during a human life", but not computer errors I guess.

Author: edited

> During the test, both buttons receive clicks. To log the errors, [test.clientScripts](https://devexpress.github.io/testcafe/documentation/reference/test-api/test/clientscripts.html) method injects a client script into the page. This script collects all the errors that happen during the test and records them into an array.

Reviewer: `To capture the errors,`

Author: done

> Read up on how to include and use custom scripts in your tests: [Inject Client Scripts](https://devexpress.github.io/testcafe/documentation/guides/advanced-guides/inject-client-scripts.html)

> The array is then checked with the [Array.some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) JavaScript method to find whether any `critical` errors have occurred on the tested web page. If such errors have occurred, the test passes, otherwise, it fails.

Reviewer 2: ```suggestion
The array is then checked with the [Array.some](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) JavaScript method to find out whether any `critical` errors have occurred on the tested web page. If such errors have occurred, then the test passes, otherwise, it fails.
```

## Launch Configuration

> To launch this example only, use the following command:

Reviewer 2: ```suggestion
To launch only this example, use the following command:
```

To launch this example only, use the following command:

```sh

> npm run extended-error-tracking;

Reviewer: It's my mistake. Need to write without semicolon.

```sh
npm run extended-error-tracking
```

Author: done
